### PR TITLE
fix(db): reject invalid and duplicate migration identifiers

### PR DESCRIPTION
**Pull request by joshtrichards**
* Resolves: # 

## Summary

This hardens migrations by preventing invalid or duplicate migration identifiers from producing an incorrect or ambiguous migration order.

Specifically:

- Migration ordering depends on the documented `<version>Date<YYYYMMDDHHMMSS>` identifier format.
- The parser accepts partial or malformed identifiers and otherwise falls back to natural string sorting.
- Recursive migration discovery can silently overwrite one file (migration) when multiple files use the same identifier across subdirectories.

Changes:

- Validate migration names against the supported version and timestamp format.
- Reject overlapping identifiers discovered in multiple migration files instead of silently overwriting one of them.
- Add test coverage for malformed filenames, stored identifiers, explicit targets, and duplicates.

## TODO

- [ ] Backport?

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)
- [x] [Labels added](https://github.com/nextcloud/server/labels) where applicable (ex: bug/enhancement, `3. to review`, feature component)
- [x] [Milestone added](https://github.com/nextcloud/server/milestones) for target branch/version (ex: 32.x for `stable32`)

## AI (if applicable)

- [x] The content of this PR was partly or fully generated using AI

### CODE REVIEW

**Review by CarlSchwan — APPROVED** [review]

**Review by provokateurin — APPROVED** [review]

File: lib/private/DB/MigrationService.php [hunk 1]
```diff
@@ -198,17 +198,32 @@ public function getAvailableVersions(): array {
...
+	}
+
+	/**
+	 * @return array{int, string}
+	 */
+	private function parseMigrationVersion(string $value): array {
+		$name = basename($value);
+
+		if (preg_match('/^Version(\d{1,16})Date(\d{14})\.php$/', $name, $matches) !== 1
+			&& preg_match('/^(\d{1,16})Date(\d{14})$/', $name, $matches) !== 1
```
Comment (come-nc) [inline lib/private/DB/MigrationService.php:218]:
```suggestion
		if (preg_match('/^(Version)?(\d{1,16})Date(\d{14})(\.php)?$/', $name, $matches) !== 1
```
Would that be harder to read?
It shifts a bit the index in $matches but otherwise should match the same things.

Can you explain why there are two supported formats though? There is only one codepath I can see.

**Comment by come-nc** [comment]
CI failures are related
```
In MigrationService.php line 220:
                                                                               
  [InvalidArgumentException]                                                   
  Invalid migration version "/home/runner/actions-runner/_work/server/server/  
  apps/cloud_federation_api/lib/Migration/Version1016Date202502262004.php" fo  
  r app "cloud_federation_api". Expected "<version>Date<YYYYMMDDHHMMSS>".

**Comment by come-nc** [comment]
And I’m not sure how this could be merged as it’s a big break for all applications which would need to rename badly named migrations, and renaming them would replay them as we rely on the name to know which one already ran.
